File: claude/web/src/lib/arcade/volley.ts
/**
 * Volley — a deterministic SteamBench arcade *temporal-vision* env.
 *
 * TypeScript mirror of `harness/steambench_harness/envs/volley.py` (and the
 * headless twin of the pixel runtime's VolleyGame). Keep a bouncing ball up: it
 * caroms off walls + ceiling and falls; slide under it to bounce it back.
 *
 * Determinism note: ALL physics is integer (positions + velocities) in the exact
 * same order as Python, so there is no float drift. The only randomness is the
 * launch — TWO draws in a fixed order: the column `rng.randrange(W - BS + 1)`,
 * then the horizontal velocity from `_LAUNCH_VX`. So `(seed, actions)` traces —
 * human, Python agent, or the vision agent reading frames — replay identically.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const W = 168;
const H = 120;
const BS = 8;
const PW = 26;
const PH = 6;
const PLAYER_Y = H - PH - 3; // 111
const PSPEED = 6;
const LAUNCH_VX = [-8, -7, 7, 8];
const VY0 = 4;
const VY_MAX = 11;
const ACTIONS = ["left", "stay", "right"] as const;

export class Volley extends Env {
  static env_id = "arcade/volley";
  static appid = 9000011;
  static displayName = "Volley";
  static description =
    "Keep the bouncing ball up: it caroms off the walls and ceiling and falls " +
    "toward the floor — slide under it to bounce it back. One miss ends the run. " +
    "A vision agent must read the ball's motion, not just its position.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace([...ACTIONS]);
  static achievements: AchievementSpec[] = [
    { id: "bounce_5", name: "Rally", description: "Bounce the ball 5 times.", rarity_hint: 0.6 },
    { id: "bounce_15", name: "Keepy-Uppy", description: "Bounce the ball 15 times.", rarity_hint: 0.3 },
    { id: "bounce_30", name: "Metronome", description: "Bounce the ball 30 times.", rarity_hint: 0.12 },
    { id: "bounce_60", name: "Wall", description: "Bounce the ball 60 times.", rarity_hint: 0.035 },
    { id: "bounce_120", name: "Unbreakable", description: "Bounce the ball 120 times.", rarity_hint: 0.006 },
  ];
  private static LADDER: [number, string][] = [
    [5, "bounce_5"], [15, "bounce_15"], [30, "bounce_30"], [60, "bounce_60"], [120, "bounce_120"],
  ];

  bx: number; by: number; vx: number; vy: number;
  px: number; bounces: number; alive: boolean;

  constructor() {
    super();
    this.bx = Math.trunc((W - BS) / 2);
    this.by = 12; this.vx = 2; this.vy = VY0;
    this.px = Math.trunc((W - PW) / 2);
    this.bounces = 0; this.alive = true;
  }

  reset(seed = 0): Observation {
    this.begin(seed);
    this.bx = this.rng.randrange(W - BS + 1);              // 1st draw
    this.by = 12;
    this.vx = LAUNCH_VX[this.rng.randrange(LAUNCH_VX.length)]; // 2nd draw
    this.vy = VY0;
    this.px = Math.trunc((W - PW) / 2);
    this.bounces = 0;
    this.alive = true;
    return this.observe(0.0);
  }

  step(action: Action): Observation {
    if (this.done) return this.observe(0.0);
    const name = this.actionSpace.name(action);
    this.steps += 1;

    if (name === "left") this.px = Math.max(0, this.px - PSPEED);
    else if (name === "right") this.px = Math.min(W - PW, this.px + PSPEED);

    this.bx += this.vx;
    this.by += this.vy;
    if (this.bx <= 0) { this.bx = 0; this.vx = -this.vx; }
    else if (this.bx >= W - BS) { this.bx = W - BS; this.vx = -this.vx; }
    if (this.by <= 0) { this.by = 0; this.vy = -this.vy; }

    let reward = 0.0;
    if (this.vy > 0 && this.by + BS >= PLAYER_Y) {
      const inBand = this.by <= PLAYER_Y + PH;
      const xOverlap = this.bx + BS > this.px && this.bx < this.px + PW;
      if (inBand && xOverlap) {
        this.bounces += 1;
        this.by = PLAYER_Y - BS;
        const mag = Math.min(VY_MAX, VY0 + Math.trunc(this.bounces / 10));
        this.vy = -mag;
        reward = 1.0;
        for (const [thresh, aid] of Volley.LADDER) {
          if (this.bounces >= thresh) this.unlock(aid);
        }
      } else if (this.by >= H) {
        this.alive = false;
        this.done = true;
        reward = -1.0;
      }
    }
    this.score = this.bounces;
    return this.observe(reward);
  }

  legalActions(): string[] {
    return [...ACTIONS];
  }

  render(): string {
    return (
      `bounces=${this.bounces} ball=(${this.bx},${this.by}) v=(${this.vx},${this.vy}) ` +
      `px=${this.px} ${this.alive ? "ALIVE" : "DEAD"}`
    );
  }

  private observe(reward: number): Observation {
    return {
      step: this.steps,
      state: {
        ball_x: this.bx, ball_y: this.by, ball_vx: this.vx, ball_vy: this.vy, ball_size: BS,
        paddle_x: this.px, paddle_y: PLAYER_Y, paddle_w: PW, paddle_h: PH, paddle_speed: PSPEED,
        width: W, height: H, bounces: this.bounces, score: this.bounces, alive: this.alive,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.bounces,
      done: this.done,
      reward: reward,
      info: { bounces: this.bounces, alive: this.alive },
    };
  }
}

File: claude/harness/steambench_harness/envs/volley.py
"""Volley — a deterministic SteamBench arcade *temporal-vision* env.

Keep a bouncing ball up: it ricochets off the walls and ceiling and falls toward
the floor; slide the paddle under it to bounce it back. Miss once and the run
ends. Unlike Dodger/Catcher (position-only perception), Volley needs *motion*:
to intercept the ball you must know its velocity, which a vision agent can only
get by differencing consecutive frames. That temporal step is the hard part of
real game-playing, and it is what the pixel runtime's CV agent demonstrates here.

All physics is INTEGER (positions + velocities), so it is bit-identical across
Python and TypeScript with no float drift; the only randomness is the ball's
initial launch (one rng draw), so ``(seed, actions)`` traces replay exactly.
"""

from __future__ import annotations

from typing import Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

# Shared with runtime/pixel_game.py VolleyGame — keep in lock-step (parity test).
W, H = 168, 120
BS = 8                  # ball size
PW, PH = 26, 6          # paddle
PLAYER_Y = H - PH - 3   # 111
PSPEED = 6              # paddle speed — deliberately SLOWER than the ball's
                        # horizontal speed, so far/fast landings can outrun it.
_LAUNCH_VX = [-8, -7, 7, 8]   # horizontal velocity (one rng draw); |vx| > PSPEED
_VY0 = 4                # initial downward speed
_VY_MAX = 11            # the rally speeds up to here (shorter reaction window)
_ACTIONS = ["left", "stay", "right"]


@register
class Volley(Env):
    env_id = "arcade/volley"
    appid = 9000011
    name = "Volley"
    description = (
        "Keep the bouncing ball up: it caroms off the walls and ceiling and "
        "falls toward the floor — slide under it to bounce it back. One miss "
        "ends the run. A vision agent must read the ball's motion, not just its "
        "position, to know where to be."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace(_ACTIONS)
    achievements = [
        AchievementSpec("bounce_5", "Rally", "Bounce the ball 5 times.", 0.6),
        AchievementSpec("bounce_15", "Keepy-Uppy", "Bounce the ball 15 times.", 0.3),
        AchievementSpec("bounce_30", "Metronome", "Bounce the ball 30 times.", 0.12),
        AchievementSpec("bounce_60", "Wall", "Bounce the ball 60 times.", 0.035),
        AchievementSpec("bounce_120", "Unbreakable", "Bounce the ball 120 times.", 0.006),
    ]
    _LADDER = [(5, "bounce_5"), (15, "bounce_15"), (30, "bounce_30"),
               (60, "bounce_60"), (120, "bounce_120")]

    def __init__(self) -> None:
        super().__init__()
        self.bx = (W - BS) // 2
        self.by = 12
        self.vx = 2
        self.vy = _VY0
        self.px = (W - PW) // 2
        self.bounces = 0
        self.alive = True

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.bx = self.rng.randrange(W - BS + 1)                  # 1st draw: launch column
        self.by = 12
        self.vx = _LAUNCH_VX[self.rng.randrange(len(_LAUNCH_VX))]  # 2nd draw: launch vx
        self.vy = _VY0
        self.px = (W - PW) // 2
        self.bounces = 0
        self.alive = True
        return self._observe(0.0)

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        name = self.action_space.name(action)
        self.steps += 1

        if name == "left":
            self.px = max(0, self.px - PSPEED)
        elif name == "right":
            self.px = min(W - PW, self.px + PSPEED)

        self.bx += self.vx
        self.by += self.vy
        # walls
        if self.bx <= 0:
            self.bx = 0
            self.vx = -self.vx
        elif self.bx >= W - BS:
            self.bx = W - BS
            self.vx = -self.vx
        # ceiling
        if self.by <= 0:
            self.by = 0
            self.vy = -self.vy

        reward = 0.0
        # Descending into the paddle's row: bounce if the paddle is under the ball
        # (ball overlaps the paddle's vertical band), else let it keep falling.
        if self.vy > 0 and self.by + BS >= PLAYER_Y:
            in_band = self.by <= PLAYER_Y + PH
            x_overlap = self.bx + BS > self.px and self.bx < self.px + PW
            if in_band and x_overlap:
                self.bounces += 1
                self.by = PLAYER_Y - BS         # sit on the paddle
                # pure vertical reflection (no english — vx only flips at walls),
                # speeding up as the rally grows (+1 every 10 bounces).
                mag = min(_VY_MAX, _VY0 + self.bounces // 10)
                self.vy = -mag
                reward = 1.0
                for thresh, aid in self._LADDER:
                    if self.bounces >= thresh:
                        self._unlock(aid)
            elif self.by >= H:                  # fell past the paddle → miss
                self.alive = False
                self.done = True
                reward = -1.0
        self.score = self.bounces
        return self._observe(reward)

    def legal_actions(self) -> list[str]:
        return list(_ACTIONS)

    def render(self) -> str:
        return (f"bounces={self.bounces} ball=({self.bx},{self.by}) v=({self.vx},{self.vy}) "
                f"px={self.px} {'ALIVE' if self.alive else 'DEAD'}")

    def _observe(self, reward: float) -> Observation:
        return Observation(
            step=self.steps,
            state={
                "ball_x": self.bx,
                "ball_y": self.by,
                "ball_vx": self.vx,
                "ball_vy": self.vy,
                "ball_size": BS,
                "paddle_x": self.px,
                "paddle_y": PLAYER_Y,
                "paddle_w": PW,
                "paddle_h": PH,
                "paddle_speed": PSPEED,
                "width": W,
                "height": H,
                "bounces": self.bounces,
                "score": self.bounces,
                "alive": self.alive,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.bounces),
            done=self.done,
            reward=float(reward),
            info={"bounces": self.bounces, "alive": self.alive},
        )
